### PR TITLE
Add multi-factor authentication

### PR DESCRIPTION
**Pull request by slavklenov**

### CODE REVIEW

File: src/mpin_sdk_base.cpp [hunk 1]
```diff
@@ -380,6 +381,46 @@ User::State User::StringToState(const String& stateString)
...
 
 /*
+* MultiFactor class
+*/
+
+MPinSDKBase::MultiFactor::MultiFactor() {}
+
+MultiFactor::MultiFactor(const char *pin) : MultiFactor(std::string(pin)) {}
+
+MPinSDKBase::MultiFactor::MultiFactor(const std::string & pin)
```
Comment (simeon-aladjem) [inline src/mpin_sdk_base.cpp]: The distractor should "zero" all the factor's strings.
I see it inherits `StringVector` - does it take care to wipe any data from the strings in the vector?

File: src/mpin_sdk_base.cpp [hunk 2]
```diff
@@ -380,6 +381,46 @@ User::State User::StringToState(const String& stateString)
...
+{
+    push_back(pin);
+}
+
+bool MPinSDKBase::MultiFactor::IsEmpty() const
+{
+    return empty() || front().empty();
+}
+
+std::string MPinSDKBase::MultiFactor::ToString() const
```
Comment (simeon-aladjem) [inline src/mpin_sdk_base.cpp]: Isn't it better to separate the factors with some delimiter?
If so, they could probably be parsed from a string, if needed

File: src/mpin_sdk_base.cpp [hunk 3]
```diff
@@ -380,6 +381,46 @@ User::State User::StringToState(const String& stateString)
...
+{
+    std::string res;
+    for (const String& f : *this)
+    {
+        res += f;
+    }
+    return res;
+}
+
+MPinSDKBase::MultiFactor::operator std::vector<int>() const
```
Comment (simeon-aladjem) [inline src/mpin_sdk_base.cpp]: The hash value here, isn't it unsigned long? I prefer to be explicit with the types ;)

File: src/mpin_sdk_base.cpp (hunk 1 above)
Comment (slavklenov) [reply src/mpin_sdk_base.cpp]: Yes, the String destructor (and thus StringVector destructor) takes care of invoking String::Ovewrite(), thus all factors Strings are automatically zeroed on MultiFactor destruction.

File: src/mpin_sdk_base.cpp (hunk 2 above)
Comment (slavklenov) [reply src/mpin_sdk_base.cpp]: MultiFactor::ToString is used only as context in unit tests. It does not matter how the concatenation will be done, as far as different factors produce different concatenated strings. If I have to add a delimiter, I have to re-generate all unit test data, but I don't see any point in this. For the reverse problem - do we actually need to specify all factors in some custom parsing format? If we really need this, it is probably better to use some existing format, like json

File: src/mpin_sdk_base.cpp (hunk 3 above)
Comment (slavklenov) [reply src/mpin_sdk_base.cpp]: The hash value is int in String::GetHash() and also pin is int in all crypto functions.

**Pull request merged**